### PR TITLE
ENH: Added Factor-VAE functionality

### PR DESCRIPTION
**Pull request by linesd**
main.py:
- Added section to prepare discriminator MLP

training.py:
- Changed input variable name from loss to loss_type for clarity
- Added an if statement to deal with the factor vae loss calculation which is dealt with separately in the losses.py file.

encoder.py:
- Changed the type of the returned latent_dist variable from tuple (with two torch.tensors) to torch.tensor. Makes dealing with it easier and makes no difference to the rest of the code.

losses.py:
- Changed the name of the latent_dist variable from tuple to torch.tensor.
- Added class FactorKLoss according to algorithm 2 of reference [1]
- Added internal function permute_dims according to algorithm 1 of reference [1]

discriminator.py:
- New discriminator MLP from reference [1] with the following architecture:
	- 6 layer mu MLP, each layer with 1000 hidden units
	- Leaky ReLu activations
	- Outputs 2 logits

References:
[1] Kim, Hyunjik, and Andriy Mnih. "Disentangling by factorising."
arXiv preprint arXiv:1802.05983 (2018).

### CODE REVIEW

File: disvae/discriminator.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
+import torch
+from torch import nn
+
+
+class Discriminator(nn.Module):
+    def __init__(self,
+                 latent_dim=10,
+                 device=torch.device("cpu")):
+        r"""Discriminator proposed in [1].
```
Comment (alecokas) [inline disvae/discriminator.py]: Does this need the string literal 'r' before the Docstring?

Comment (linesd) [reply disvae/discriminator.py]: I don't use any backslashes in the docstring so the 'r' is not necessary.  Happy to remove it.

Comment (alecokas) [reply disvae/discriminator.py]: I don't feel strongly about it at all, I was more interested if you had a reason for doing it.

File: disvae/losses.py [hunk 2]
```diff
@@ -167,6 +169,100 @@ def __call__(self, data, recon_data, latent_dist, is_train, storer):
...
+        Compute the Factor-VAE loss as per Algorithm 2 of [1]
+
+        Parameters
+        ----------
+        is_color : bool
+            Whether the image are in color.
+
+        discriminator : disvae.discriminator.Discriminator
+
+        optimizer_d : torch.optim.adam.Adam
```
Comment (YannDubs) [inline disvae/losses.py]: extremely minor note but this is is not restricted to Adam (i.e. `torch.optim`)

File: disvae/losses.py [hunk 3]
```diff
@@ -167,6 +169,100 @@ def __call__(self, data, recon_data, latent_dist, is_train, storer):
...
+        self.optimizer_d = optimizer_d
+
+    def __call__(self, data, model, optimizer, is_train, storer):
+        storer = self._pre_call(is_train, storer)
+
+        # If factor-vae split data into two batches
+        batch_size = data.size(dim=0)
+        half_batch_size = batch_size // 2
+        data = data.split(half_batch_size)
+        data1 = data[0].to(self.device)
```
Comment (YannDubs) [inline disvae/losses.py]: The data should already be on the correct device. 

As a side note, even if it weren't, you would not put it on GPU for each batch. Indeed CPU -> GPU takes time, so better do it once (as done now) than for each batch. 

Note that currently it should nto change anything because i'm pretty sure pytorch checks whether it's already on the correct device before moving it :)

**Comment by linesd** [comment]
Thanks for these comments. Both really nice points. Please continue to comment on my code even for the smallest subtleties. A nice way to learn and correct small things when learning new libraries!

File: disvae/losses.py [hunk 4]
```diff
@@ -229,3 +325,32 @@ def _kl_normal_loss(mean, logvar, storer=None):
...
+        sufficient statistics of the latent dimension. E.g. for gaussian
+        (mean, log_var), each of shape : (batch_size, latent_dim).
+
+    References :
+        [1] Kim, Hyunjik, and Andriy Mnih. "Disentangling by factorising."
+        arXiv preprint arXiv:1802.05983 (2018).
+
+    """
+
+    perm = torch.zeros(latent_dist.size())
```
Comment (YannDubs) [inline disvae/losses.py]: did you test on GPU? suprised it works! It seem to me that perm is still on cpu. YOu should do `torch.zeros_like(latent_dist)`

File: disvae/encoder.py [hunk 5]
```diff
@@ -89,4 +89,4 @@ def forward(self, x):
         mu = self.mu_gen(x)
         log_var = self.log_var_gen(x)  # Log std-dev in paper (bear in mind)
 
-        return mu, log_var
+        return torch.stack([mu, log_var])
```
Comment (YannDubs) [inline disvae/encoder.py]: in general in Pytorch we alway leave dimesnion 1 to be batch size. Stacking is totally fine (and more general), but if you want to be clean I'd stack on `dim=-1` (i.e. last dimesnion)

File: disvae/losses.py [hunk 6]
```diff
@@ -229,3 +325,32 @@ def _kl_normal_loss(mean, logvar, storer=None):
...
+    dim_j, batch_size, dim_z = latent_dist.size()
+
+    for j in range(dim_j):
+        for z in range(dim_z):
+            pi = torch.randperm(batch_size)
```
Comment (YannDubs) [inline disvae/losses.py]: not sure if indices have to be on CUDA, if yes then you are missing `device=device`

File: disvae/losses.py [hunk 7]
```diff
@@ -229,3 +325,32 @@ def _kl_normal_loss(mean, logvar, storer=None):
...
+        [1] Kim, Hyunjik, and Andriy Mnih. "Disentangling by factorising."
+        arXiv preprint arXiv:1802.05983 (2018).
+
+    """
+
+    perm = torch.zeros(latent_dist.size())
+    dim_j, batch_size, dim_z = latent_dist.size()
+
+    for j in range(dim_j):
+        for z in range(dim_z):
```
Comment (YannDubs) [inline disvae/losses.py]: IMPORTANT : you seem to be permuting the latent_dist (i.e mean and variance) while in the paper (and in 1konny) they permute the samples!!!!!!! (that's why you have one more loop.
I'm pretty sure this is not the same. One thing that might be the same (i haven't though about it too much yet) is if you use the same permutation for the mean and the variance. But it also shows a fudamental difference between you code and the paper: you are actually giving the distribution to the discriminator not z !! FOr example you say `z_perm = _permute_dims(latent_dist)` but it's actually `latent_dist_perm = _permute_dims(latent_dist)`. I'd have to think abou tit more carefully if that is also correct.
Are you able to get good results?

File: disvae/losses.py [hunk 8]
```diff
@@ -2,12 +2,11 @@
 Module containing all vae losses.
 """
 import abc
-
 import torch
 from torch.nn import functional as F
 
 
-def get_loss_f(name, is_color, capacity=None, beta=None):
+def get_loss_f(name, is_color, capacity=None, beta=None, discriminator=None, optimizer_d=None, device=None):
```
[the remark this answers is not in the document]
Comment (linesd) [reply disvae/losses.py]: Will do

File: disvae/losses.py (hunk 4 above)
Comment (linesd) [reply disvae/losses.py]: Good point, I'll change this

File: disvae/encoder.py (hunk 5 above)
Comment (linesd) [reply disvae/encoder.py]: Again, nice point, I'll change this

File: disvae/losses.py [hunk 9]
```diff
@@ -167,6 +169,100 @@ def __call__(self, data, recon_data, latent_dist, is_train, storer):
...
+        rec_loss = _reconstruction_loss(data1, recon_batch, self.is_color)
+
+        # Get KL-Divergence (latent_dist[0] = mean, latent_dist[1] = log_var)
+        kl_loss = _kl_normal_loss(latent_dist[0], latent_dist[1], storer)
+
+        # Run latent distribution through discriminator
+        d_z = self.discriminator(latent_dist)
+
+        # Calculate the total correlation (TC) loss term
+        tc_loss = (d_z[:, :, :1] - d_z[:, :, 1:]).mean()
```
[the remark this answers is not in the document]
Comment (linesd) [reply disvae/losses.py]: When I first read the paper I interpreted it similarly to you (the logistic regression one). But then looking at 1konny's code (and given he/she gets good results)  I thought I had misinterpreted it. Your thinking is correct that the implementation is doing log(D(z))-log(1-D(z)) but you make a very good point about the constraint. Lets have a chat about it tomorrow.

File: disvae/losses.py (hunk 6 above)
Comment (linesd) [reply disvae/losses.py]: Not sure what you mean here by `device=device`? Do you mean it should be `pi = torch.randperm(batch_size).to(latent_dist.device)`?

File: disvae/losses.py (hunk 2 above)
Comment (linesd) [reply disvae/losses.py]: I'll change this.

File: disvae/losses.py (hunk 3 above)
Comment (linesd) [reply disvae/losses.py]: Thanks, I'll remove it

File: disvae/discriminator.py [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+        """
+        super(Discriminator, self).__init__()
+
+        # Activation parameters
+        neg_slope = 0.2
+        self.leaky_relu = nn.LeakyReLU(neg_slope, True)
+
+        # Layer parameters
+        self.z_dim = latent_dim
+        hidden_units = 1000
```
[the remark this answers is not in the document]
Comment (linesd) [reply disvae/discriminator.py]: Agreed, I'll change it.

File: main.py [hunk 11]
```diff
@@ -177,15 +178,22 @@ def main(args):
     decoder = get_Decoder(args.model)
     model = VAE(img_size, encoder, decoder, args.latent_dim, device=device)
 
+    # PREPARES DISCRIMINATOR FOR FACTOR VAE
```
[the remark this answers is not in the document]
Comment (linesd) [reply main.py]: Thanks for the tip. I agree with the recommendation. I'll change this in the code too.

File: disvae/losses.py (hunk 7 above)
Comment (linesd) [reply disvae/losses.py]: Thanks for pointing this out. I'll fix the error!

File: disvae/losses.py (hunk 6 above)
Comment (YannDubs) [reply disvae/losses.py]: yes might be (not sure whether indices have to be on CUDA, but I guess so)
And I was saying `torch.randperm(batch_size, device=latent_dist.device)` because if you use `.to()` you first allocate on CPU then have to move it to GPU

File: disvae/losses.py (hunk 3 above)
Comment (linesd) [reply disvae/losses.py]: Yann and I agreed offline that the .to(self.device) should be included here.

Comment (YannDubs) [reply disvae/losses.py]: check the current master branch (due to my last PR) I actually call `.to(self.device)` before the if statement

Comment (linesd) [reply disvae/losses.py]: Okay, great. Any splitting the data doesn't affect this?

Comment (YannDubs) [reply disvae/losses.py]: no

Comment (linesd) [reply disvae/losses.py]: Fair enough. The only reason I changed it was for the loop which now won't be happening.

File: disvae/losses.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (linesd) [reply disvae/losses.py]: I took another look at the paper and they say in Appendix A: 

> We use the same 6 layer MLP discriminator with 1000 hidden units per layer and leaky ReLU (lReLU) non-linearity, that outputs `2 logits` in all FactorVAE experiments.

So I think the discriminator implementation is actually correct. I'm wondering if the constraint is actually enforced by this line of code?:

`d_tc_loss = 0.5 * (F.cross_entropy(d_z, zeros) + F.cross_entropy(d_z_perm, ones))`
